File: MusicPlayer/app/src/main/java/com/zukron/musicplayer/activity/MainActivity.java
package com.zukron.musicplayer.activity;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.recyclerview.widget.RecyclerView;

import android.Manifest;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.View;

import com.codekidlabs.storagechooser.StorageChooser;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.zukron.musicplayer.Preferences;
import com.zukron.musicplayer.R;
import com.zukron.musicplayer.adapter.LibraryAdapter;
import com.zukron.musicplayer.model.Library;

import java.io.File;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

public class MainActivity extends AppCompatActivity implements View.OnClickListener, StorageChooser.OnSelectListener, LibraryAdapter.OnSelected {
    private RecyclerView rvMain;
    private FloatingActionButton fabMain;
    private StorageChooser storageChooser;
    private ArrayList<Library> libraries;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        rvMain = findViewById(R.id.rv_main);
        fabMain = findViewById(R.id.fab_main);

        loadPreferences();
        permission();
        init();
    }

    private void loadPreferences() {
        libraries = Preferences.loadPreferences(getSharedPreferences(Preferences.NAME, Preferences.MODE_PRIVATE));

        if (libraries == null) {
            libraries = new ArrayList<>();
        } else {
            setRecyclerView();
        }
    }

    private void permission() {
        if (checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, PackageManager.PERMISSION_GRANTED);
        }
    }

    private void init() {
        // theme storage chooser
        StorageChooser.Theme theme = new StorageChooser.Theme(this);
        theme.setScheme(getResources().getIntArray(R.array.paranoid_theme));

        // storage chooser builder
        storageChooser = new StorageChooser.Builder()
                .withActivity(this)
                .setTheme(theme)
                .withFragmentManager(getFragmentManager())
                .withMemoryBar(true)
                .allowCustomPath(true)
                .setType(StorageChooser.DIRECTORY_CHOOSER)
                .build();

        fabMain.setOnClickListener(this);
        storageChooser.setOnSelectListener(this);
    }

    private void setRecyclerView() {
        LibraryAdapter libraryAdapter = new LibraryAdapter(this, libraries);
        libraryAdapter.setOnSelected(this);
        rvMain.setAdapter(libraryAdapter);
    }

    @Override
    public void onClick(View view) {
        storageChooser.show();
    }

    @Override
    public void onSelect(String path) {
        File file = new File(path);

        if (file.exists() || file.canRead()) {
            if (isFolderContainsMp3(file)) {
                libraries.add(new Library(file.getName(), file.getAbsolutePath()));
                removeDuplicate(libraries);

                setRecyclerView();
                Preferences.savePreferences(libraries, getSharedPreferences(Preferences.NAME, Preferences.MODE_PRIVATE));
            } else {
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle("Failed");
                builder.setMessage("File .mp3 tidak terdeteksi");
                builder.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        dialogInterface.dismiss();
                    }
                });

                builder.show();
            }
        }
    }

    private boolean isFolderContainsMp3(File file) {
        boolean valid = false;
        File[] files = file.listFiles();

        if (files != null) {
            for (File value : files) {
                valid = value.getAbsolutePath().contains(".mp3");
                break;
            }
        }

        return valid;
    }

    private void removeDuplicate(ArrayList<Library> libraries) {
        Set<Library> set = new HashSet<>(libraries);
        libraries.clear();
        libraries.addAll(set);
    }

    @Override
    public void onSelectedItem(Library library) {
        Intent intent = new Intent(this, DetailMusicActivity.class);
        intent.putExtra("Library", library);
        startActivity(intent);
    }
}

File: MusicPlayer/app/src/main/java/com/zukron/musicplayer/activity/MainActivity.kt
package com.zukron.musicplayer.activity

import android.Manifest
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.codekidlabs.storagechooser.StorageChooser
import com.zukron.musicplayer.MusicPreferences
import com.zukron.musicplayer.R
import com.zukron.musicplayer.adapter.LibraryAdapter
import com.zukron.musicplayer.model.Library
import kotlinx.android.synthetic.main.activity_main.*
import java.io.File

/**
 * Project name is Music Player
 * Created by Zukron Alviandy R on 9/3/2020
 * Contact me if any issues on zukronalviandy@gmail.com
 */
class MainActivity : AppCompatActivity() {

    private var libraries = mutableListOf<Library>()
    private val libraryAdapter = LibraryAdapter(this)
    private var storageChooser: StorageChooser? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // check permission
        if (checkSelfPermission(
                        Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), PackageManager.PERMISSION_GRANTED)
        }

        // load preferences
        val librariesPref = MusicPreferences.load(this)
        librariesPref.let {
            libraries = it as MutableList<Library>
            libraryAdapter.libraries = libraries

            recycler_main.adapter = libraryAdapter
        }

        // storage chooser builder
        val storageChooserTheme = StorageChooser.Theme(this)
        storageChooserTheme.scheme = resources.getIntArray(R.array.paranoid_theme)

        storageChooser = StorageChooser.Builder()
                .withActivity(this)
                .setTheme(storageChooserTheme)
                .withFragmentManager(fragmentManager)
                .withMemoryBar(true)
                .allowCustomPath(true)
                .setType(StorageChooser.DIRECTORY_CHOOSER)
                .build()
    }

    override fun onStart() {
        super.onStart()

        libraryAdapter.onSelected = object : LibraryAdapter.OnSelected {
            override fun onSelectedItem(library: Library) {
                val intent = Intent(this@MainActivity, DetailMusicActivity::class.java)
                intent.putExtra(DetailMusicActivity.EXTRA_LIBRARY, library)
                startActivity(intent)
            }
        }

        btn_floating_action.setOnClickListener {
            storageChooser?.show()
        }

        storageChooser?.setOnSelectListener {
            val file = File(it)
            val isContainsMp3: () -> Boolean = {
                var valid = false
                val fileList: Array<File>? = file.listFiles()
                if (fileList != null) {
                    for (fileVal in fileList) {
                        val isMp3 = fileVal.absolutePath.contains(".mp3").also { exist ->
                            valid = exist
                        }
                        if (isMp3) break
                    }
                }
                valid
            }

            if (file.exists() || file.canRead()) {
                val exist = isContainsMp3()
                if (exist) {
                    libraries.add(Library(file.name, file.absolutePath))
                    libraries = libraries.distinct().toMutableList()

                    libraryAdapter.libraries = libraries
                    libraryAdapter.notifyDataSetChanged()

                    MusicPreferences.save(libraries, this)
                } else {
                    val alertDialogBuilder = AlertDialog.Builder(this)
                    alertDialogBuilder.setTitle("Failed")
                    alertDialogBuilder.setMessage("File .mp3 tidak terdeteksi")
                    alertDialogBuilder.setPositiveButton("Ok", object : DialogInterface.OnClickListener {
                        override fun onClick(dialogInterface: DialogInterface?, p1: Int) {
                            dialogInterface?.dismiss()
                        }
                    })

                    alertDialogBuilder.show()
                }
            }
        }
    }
}
